### PR TITLE
Add built-in user and role for code plugin

### PR DESCRIPTION
**Pull request by spacedragon**
This pr add a `code_system` as a   reserved role, and grant index privilege `.code-*` to both `code_system` and `kibana_system`.

### CODE REVIEW

**Comment by zfy0701** [comment]
@spalger @polyfractal could you help reviewing this PR?

**Comment by tvernum** [comment]
Why are we creating a new role, but not a new user? How will the `code_system` role be used?

**Comment by spacedragon** [comment]
@tvernum By default, kibana users have sufficient permissions to manage code plugins, but if a customer wants to grant other users permission to manage code, they can assign code_system to them.

**Comment by polyfractal** [comment]
@tvernum would you mind taking the review on this?  Not sure I know enough about security to intelligently review past syntax/style :)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (mw-ding) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: @spacedragon I assume this role is used for non Code admin user, who will only have READ permission to code indices. And to allow a user to have READ-ONLY permission, the admin of Kibana has to grant `kibana_user` role to the user, right? Here the `kibana_user` will have `all` permission to `.code-` indices though.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (mw-ding) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: nit: add the settings for `.code-*` indices separately here to make it more readable?

e.g. adding
```
.indices(".code-*").privileges("create_index", "read", "write").build(),
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (legrego) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: The `kibana_user` role is slowly becoming more of a "Kibana administrator". Users with the `kibana_user` role have full read/write access to all spaces, so I think it makes sense in this context to have full access to the code indices. I don't know a whole lot about Code, so let me know if that sounds wrong.

Comment (mw-ding) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: @legrego Yea, this makes sense. In our case, we probably need to have a role with only read permission as a *regular* Code user as opposed to an *admin* Code user.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (spacedragon) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: done

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (spacedragon) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: changed `code_system` to `code_admin`

**Comment by tvernum** [comment]
I will review, but I'll be out for the next couple of days, so I won't get to it until mid next week.

It would be helpful if you could provide some background about how you intend for this to work. I can review the change at a purely technical level, but in order to understand whether it's the _right_ change to be making I'll need to understand what these `.code*` indices are used for and which users/applications read/write to them.  
At the moment I'm confused by the fact that we're changing the `kibana_system` user (which the Kibana backend uses) and creating a `code_admin` user (which seems to be intended for real people). They seem very different and I'm not sure how they fit together.

Also, before I can approve this, we'll need corresponding changes to `ReservedRolesStoreTests`.

**Comment by zfy0701** [comment]
so we are trying to mimic how beats/apm security works in Kibana

first, the kibana_system role would have read/write access to all code-* indices because it's the root role

then, for users that are not root user but need to have permission to manage code, they need to be granted code_admin role, this is similar to beats_admin I assume

last, for normal users, they need to be granted read access to all code-* indices, we initially planned to create a code_user role, but it's seems beats and apm don't do that, so we just follow the pattern.

Please let us known if we have misunderstood how beats/apm works in Kibana, it would be super helpful

**Comment by tvernum** [comment]
Thanks @spacedragon.
I'm happy with the change - we just need to update `ReservedRolesStoreTests` to have tests for these 2 new roles that mimic the existing tests. It should be self-explanatory, but ping me if you need more info.

**Comment by tvernum** [comment]
Those failures don't look related.

@elasticmachine run gradle build tests 1 ; run gradle build tests 2

**Comment by tvernum** [comment]
@spacedragon I fixed a couple of tests, we'll see how CI goes.

**Comment by tvernum** [comment]
@elasticmachine test this please

**Comment by tvernum** [comment]
Hey @elasticmachine
run elasticsearch-ci-1
run elasticsearch-ci-2

**Comment by tvernum** [comment]
Hey @elasticmachine
run elasticsearch-ci/1
run elasticsearch-ci/2

**Comment by spacedragon** [comment]
@tvernum Thanks a lot for your help, could I merge this pr now?

**Comment by tvernum** [comment]
@spacedragon We need to get the build green before we can merge. I haven't been checking the most recent failures because Elasticsearch CI has had a bunch of internal problems, but they _should_ be resolved now.
If this build fails we'll need to look at why that's happening.

**Comment by spacedragon** [comment]
It's all green now.

**Comment by tvernum** [comment]
@spacedragon Feel free to merge this.

**Review by tvernum — APPROVED** [review]

**Comment by tvernum** [comment]
@spacedragon I've labelled this on the assumption that you will want to backport this to 6.7
Let me know if you need assistance with that.

**Comment by spacedragon** [comment]
@tvernum Thanks.
@zfy0701 What's the requirement of elasticsearch for our first code release? I think it's 7.0.0, right?

**Comment by tvernum** [comment]
I dropped the `6.7.0` label since it didn't get backported. Hopefully you only wanted this on 7.0

**Comment by legrego** [comment]
I would like to _partially_ revert this PR. With the introduction of Kibana Feature Controls ([and Code's support for it](https://github.com/elastic/kibana/pull/35115)), the reserved `code_admin` and `code_user` roles are no longer necessary. 
Is is possible to remove reserve roles in a minor release?

As far as I can tell, those roles aren't doing anything for customers today, as the corresponding code app isn't available yet. So if we remove this role, we shouldn't be revoking any access that users are relying on today.

@elastic/codesearch do you agree, or am I overlooking anything here?

**Comment by zfy0701** [comment]
> I would like to _partially_ revert this PR. With the introduction of Kibana Feature Controls ([and Code's support for it](https://github.com/elastic/kibana/pull/35115)), the reserved `code_admin` and `code_user` roles are no longer necessary.
> Is is possible to remove reserve roles in a minor release?
> 
> As far as I can tell, those roles aren't doing anything for customers today, as the corresponding code app isn't available yet. So if we remove this role, we shouldn't be revoking any access that users are relying on today.
> 
> @elastic/codesearch do you agree, or am I overlooking anything here?

agree, I don't think we need them

**Pull request merged**
